GridKit/Model/PowerElectronics/SystemModelPowerElectronics.hpp

```diff
@@ -121,20 +126,177 @@ namespace GridKit
+            continue;
+
+          global_row_indices[curr_internal_row + 1] = global_row_indices[curr_internal_row];
+
+          assert(component->getNodeConnection(local_row) == curr_internal_row);
+
+          IdxT row_start = comp_sparsity.row_indices_[local_row];
+          IdxT row_end   = comp_sparsity.row_indices_[local_row + 1];
+          for (IdxT local_col_idx = row_start; local_col_idx < row_end; local_col_idx++)
+          {
+            IdxT global_col_idx = component->getNodeConnection(comp_sparsity.col_indices_[local_col_idx]);
+
+            if (global_col_idx == neg1_)
+              continue;
+
+            global_col_indices[global_row_indices[curr_internal_row + 1]] = global_col_idx;
+            global_row_indices[curr_internal_row + 1]++;
+          }
+
+          // Need to sort internal rows because even though the mapping from component internal to system internal
+          // is monotonic, the mapping from component external to system external may not be, and internal rows
+          // may contain external columns.
+          auto global_row_start = global_col_indices + global_row_indices[curr_internal_row];
+          auto global_row_end   = global_col_indices + global_row_indices[curr_internal_row + 1];
+          std::sort(global_row_start, global_row_end);
+
+          curr_internal_row++;
+        }
+      }
+
+      assert(curr_internal_row == n_intern_);
+
+      // Then move on to external rows, which come after internal rows
+      for (size_t row = n_intern_; row < size_; row++)
+      {
+        global_row_indices[row + 1] = global_row_indices[row];
+
+        // Collect columns from each component which has a row which contributes to this row
+        for (ComponentContribution contrib : reverse_extern_map[row - n_intern_])
+        {
+          component_type* component     = components_[contrib.comp_idx_];
+          CsrSparsity&    comp_sparsity = component_sparsities[contrib.comp_idx_];
+
+          IdxT row_start = comp_sparsity.row_indices_[contrib.local_row_idx_];
+          IdxT row_end   = comp_sparsity.row_indices_[contrib.local_row_idx_ + 1];
+          for (size_t local_elem_idx = row_start; local_elem_idx < row_end; local_elem_idx++)
+          {
+            IdxT local_col  = comp_sparsity.col_indices_[local_elem_idx];
+            IdxT global_col = component->getNodeConnection(local_col);
+
+            // Not all columns map to columns in the system jacobian
+            if (global_col != neg1_)
+            {
+              global_col_indices[global_row_indices[row + 1]] = global_col;
+              global_row_indices[row + 1]++;
+            }
+          }
+        }
+
+        // Sort the row by column indices. Since the mapping from local indices to global indices isn't monotonically increasing,
+        // this is necessary.
+        auto start = std::next(global_col_indices, static_cast<ptrdiff_t>(global_row_indices[row]));
```

Comment:
I added a `static_cast<ptrdiff_t` to get around sign warnings from Clang. @alexander-novo can we avoid this iterator altogether?